### PR TITLE
Initial PR setting up repo structure

### PR DESCRIPTION
**Pull request by MarieStLaurent**
Debugged code extracts denoised timeseries from cneuromod dataset, recombining datasets, parcellations and denoising strategies with hydra config files.

Files to review include the .py scripts found here (run.py, extract.py and utils.py): https://github.com/courtois-neuromod/cneuromod_extract_tseries/tree/dev_clean/timeseries

.yaml config files are here : https://github.com/courtois-neuromod/cneuromod_extract_tseries/tree/dev_clean/timeseries/config

And links to the doc are found here: https://github.com/courtois-neuromod/cneuromod_extract_tseries/blob/dev_clean/docs/overview.rst

### CODE REVIEW

**Comment by MarieStLaurent** [comment]
@htwangtw if you could review that would be great!

File: timeseries/utils.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+
+PRESET_STRATEGIES = [
+    "simple",
+    "simple+gsr",
+    "scrubbing.2",
+    "scrubbing.2+gsr",
+    "scrubbing.5",
+    "scrubbing.5+gsr",
+    "acompcor50",
+    "icaaroma",
```
Comment (htwangtw) [inline timeseries/utils.py:40]: We are not shipping ICA-AROMA

File: timeseries/utils.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+        raise ValueError("No valid subject directory found.")
+
+    return checked_labels
+
+
+def parse_standardize_options(
+    standardize: str,
+) -> Union[str, bool]:
+    # TODO: update standardization choices based on nilearn warnings
+    if standardize not in ["zscore_sample", "psc"]:
```
Comment (htwangtw) [inline timeseries/utils.py:95]: I recommend to simplify this by only outputting zscore. PSC, after some deep dive, is a very weird thing outside of the task contrast context IMO. We can remove this function entirely.

File: timeseries/utils.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        The masked time series data array.
+    """
+    time_series_atlas = masker.fit_transform(denoised_img)
+
+    # convert to float 32 instead of 64
+    return time_series_atlas.astype(np.float32)
+
+
+
+def _get_consistent_masks(
```
Comment (htwangtw) [inline timeseries/utils.py:159]: This is not relevant if we are using individual space in cneuromod

File: timeseries/utils.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+    standardize: str,
+) -> Union[str, bool]:
+    # TODO: update standardization choices based on nilearn warnings
+    if standardize not in ["zscore_sample", "psc"]:
+        raise ValueError(f"{standardize} is not a valid standardize strategy.")
+
+    return standardize
+
+
+def prep_denoise_strategy(
```
Comment (htwangtw) [inline timeseries/utils.py:101]: avoid function this small but with extensive docs.

File: timeseries/utils.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+    dict
+        Denosing strategy parameter to pass to load_confounds_strategy.
+    """
+    lc_function = getattr(fmriprep, benchmark_strategy["function"])
+    benchmark_strategy.update({"function": lc_function})
+
+    return benchmark_strategy
+
+
+def generate_timeseries(
```
Comment (htwangtw) [inline timeseries/utils.py:133]: see line 101

File: timeseries/utils.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+    odd_masks = odd_masks.tolist()
+    for odd_file in odd_masks:
+        identifier = Path(odd_file).name.split("_space")[0]
+        weird_mask_identifiers.append(identifier)
+    cleaned_func_masks = set(mask_imgs) - set(odd_masks)
+    cleaned_func_masks = list(cleaned_func_masks)
+    return cleaned_func_masks, weird_mask_identifiers
+
+
+def _check_mask_affine(
```
Comment (htwangtw) [inline timeseries/utils.py:193]: possibly not relevant

File: timeseries/utils.py [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+        )
+        exclude += ob_index
+    gc_log.info(
+        f"{len(exclude)} out of {len(mask_imgs)} has "
+        "different affine matrix. Ignore when creating group mask."
+    )
+    return sorted(exclude)
+
+
+def merge_masks(
```
Comment (htwangtw) [inline timeseries/utils.py:261]: if we want to do masking with grey matter mask, this should be adapted accordingly

File: timeseries/utils.py [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+    Nifti1Image
+        Denoised nifti image.
+    """
+    cf, sm = strategy["function"](img, **strategy["parameters"])
+    if _check_exclusion(cf, sm):
+        return None
+
+    # if high pass filter is not applied through cosines regressors,
+    # then detrend
+    detrend = "cosine00" not in cf.columns
```
Comment (htwangtw) [inline timeseries/utils.py:330]: Remove detrend check for cneuromod project for simplicity. we are always using fMRIPrep cosine regressors

File: timeseries/utils.py [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+    subject_masker = NiftiMasker(
+        mask_img=subject_mask,
+        detrend=detrend,
```
Comment (htwangtw) [inline timeseries/utils.py:333]:
```suggestion

```

File: timeseries/utils.py (hunk 8 above)
Comment (htwangtw) [inline timeseries/utils.py:330]:
```suggestion
```

File: timeseries/run.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import hydra
+from omegaconf import DictConfig, OmegaConf
+
+
+@hydra.main(
```
Comment (htwangtw) [inline timeseries/run.py:5]: After looking at the full code base, I suggest to remove all hydra components. It's bulky, and adds future maintenance that's not really needed(such as add dataset name to the repository)

File: timeseries/extract.py [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+from nilearn.maskers import NiftiLabelsMasker, NiftiMapsMasker, NiftiMasker
+from nilearn.masking import compute_multi_epi_mask
+import numpy as np
+from omegaconf import DictConfig
+from tqdm import tqdm
+
+from timeseries import utils
+
+
+class ExtractionAnalysis:
```
Comment (htwangtw) [inline timeseries/extract.py:22]: what's the motivation to make this a class?
if things can be resolved with functions, use functions

File: timeseries/utils.py [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+    subject_mask_nii = math_img(
+        "img1 & img2",
+        img1=subject_epi_mask,
+        img2=mni_gm_mask_nii,
+    )
+
+    return subject_mask_nii
+
+
+def denoise_nifti_voxel(
```
Comment (htwangtw) [inline timeseries/utils.py:295]: some small functions i pointed out above can be merged here

**Review by htwangtw — COMMENTED** [review]
I am focusing on high level comments and a few changes related the details of denoising. I skip the atlas and masker part because this is too much to unpack. This can be downsized through redesign. At this stage is more useful to have a redesign than detailed review.

## Denosing:

For the denoising strategies, if we are focused on time series analysis, all the scrubbing based strategy can be dropped. This will need to be discussed with the team. 

## Mask / atlas:

I don't have the capacity to figure what's going on in this PR with a class wrapping around what's going on.

## Code:

I was expecting to see a clone of `giga_connectome` with heavy revision on the mask module on how to handle atlases mostly, but that part is now part of a class of analysis? I really don't see a reason to use a class here if this can be handled by a function.

Some functions became two lines since here we are expecting to lock-in the decisions on denoising methods, and they should just be merged with other parts. 
For example: 
- the denoise module  of `giga_connectome` would just be come one function here.
- the connectome module  of `giga_connectome` will be complete gone but just two lines because there's no extra metrics to calculate here.  

As for hydra components, it's adding too many files for things we don't want to expose to users to change on the fly. 
  - datasets can just be provided from user side.
  - we don't really want to have the ability to change how each denoising method is executed so I would not expose them to users. 
I would really recommand to use `argparse`.

As for moving forward, an easier work flow for review and collaboration would be:
- to have some basic code, either this one or a fresh fork from `giga_connectome`
- open issues for each feature we want to change and implement each as a PR. We can use the documentation and points raised here for guidance, and I am happy to make sure the size of the task is a reasonable size for PRs. 

I would prefer to start from `giga_connectome` since that can potentially benefit `giga_connectome` if we can find better solutions than the current code base.
